core/src/main/scala/kafka/log/Log.scala

```diff
@@ -497,8 +574,12 @@ class Log(@volatile var dir: File,
     var monotonic = true
     var maxTimestamp = RecordBatch.NO_TIMESTAMP
     var offsetOfMaxTimestamp = -1L
+    var isDuplicate = false
+    val producerAppendInfos = mutable.Map[Long, ProducerAppendInfo]()
 
     for (batch <- records.batches.asScala) {
+      if (isFromClient && batch.magic >= RecordBatch.MAGIC_VALUE_V2 && shallowMessageCount > 0)
+        throw new InvalidRecordException("Client produce requests should not have more than one batch")
```

Comment:
We did, initially, but then I followed up on that thread (which is now impossible to find on github). Here is what I wrote: 

> So I thought about this a bit more. I think it still make sense to validate this on the server side. I assume that the librdkafka clients may not have the request side validation, so it would be good to have server side validation before we write to the log.